### DISCUSSION TITLE
Include possible HTTPExceptions in OpenAPI spec

### DISCUSSION THREAD

**Discussion by onecrayon** [fastapi/fastapi#9124] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, HTTPException

app = FastAPI()


@app.get("/")
def read_root(gimme_coffee: bool = False):
    if gimme_coffee:
        raise HTTPException(status_code=418, detail="I'm a teapot.")
    return {"Hello": "World"}
```

### Description

* Open the browser and call the endpoint `/docs`
* Note that the only listed response codes are 200 and 422
* Execute the route with `gimme_coffee` set to true and note that it returns a 418 status code

How do I set the status code for 418 in the OpenAPI docs? I was expecting to be able to pass in a list of possible HTTPExceptions and have them be automatically converted, but there does not appear to be any way to do that. The closest thing I found was [Additional Status Codes](https://fastapi.tiangolo.com/advanced/additional-status-codes/) but that seemed at best tangentially relevant since I am raising HTTPExceptions, not looking to build out a custom exception JSON response.

Properly documenting error states is just as important as documenting the success state. I really hope FastAPI provides an easy way to do this.

### Environment

* OS: Linux (Docker container using standard python-3.8 image)
* FastAPI Version: 0.61.1
* Python version: 3.8.5

**Comment by ycd** [comment]
Yes, you can declare additional responses with additional status codes etc. You just need to add it as a path parameter then those additional responses will be included in the OpenAPI schema, here is a example

```python
@app.get("/",responses={418: {"detail": "I'm a teapot"}, 413: {"Too large": "...The Payload"}})
def read_root(gimme_coffee: bool = False):
    if gimme_coffee:
        raise HTTPException(status_code=418, detail="I'm a teapot.")
    return {"Hello": "World"}
```

Documentation for  [Additional Responses](https://fastapi.tiangolo.com/advanced/additional-responses/)

**Comment by onecrayon** [comment]
That's the additional status codes I mentioned above, and it is a long way from easy. I was hoping I could instantiate an error, pass it in similar to `response_model` and have the OpenAPI generation just do the right thing, but looks like I would need to basically duplicate the entire error into the decorator.

Nuts. I guess this is actually a feature request: we need a better way to handle error message documentation. My preference would be to just pass in a list of some base error classes and have it figure out the status code and output formatting based on that (having the actual error output isn't necessary, particularly since it may well change depending on the situation). That way, I could setup a sub-class of HTTPException that defaults to a 418 status code, and by passing that in and using it my exception is fully documented.

**Comment by David-Lor** [comment]
@onecrayon On my APIs I'm using custom exceptions that include references to the error code, message and Pydantic model to be returned. I have an example over here: https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API

On a nutshell:

- I create the custom exceptions (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/exceptions.py)
- I create the models for exception responses (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/models/errors.py)
- On the routes, I call a function that creates the "responses" dict, given multiple custom exceptions (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/app.py)
- I use a request middleware, where a try/except can catch my custom exceptions, and return their custom response, based on the exception status code, message and model (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/middlewares.py)

Probably far from ideal and might be better ways to achieve it, but for now it's been working pretty fine for me. It's more complex though.

**Comment by ycd** [comment]
@onecrayon Does something like this would work for you?`

```python
from fastapi import FastAPI, APIRouter


app = FastAPI()
router = APIRouter()


@router.get("/dummy1")
async def dummy1():
    ...


@router.get("/dummy2")
async def dummy2():
    ...


@router.get("/dummy3")
async def dummy3():
    ...


@router.get("/dummy4")
async def dummy4():
    ...


app.include_router(
    router,
    responses={
        418: {
            "description": "I'm a teapot",
            "content": {"application/json": {"example": {"dummy": "dumdum"}}},
        },
        413: {
            "description": "I'm too large",
            "content": {"application/json": {"example": {"dummy": "dumdum"}}},
        },
    },
)
```



![image](https://user-images.githubusercontent.com/62724709/100586198-d6febc80-32ff-11eb-9bae-37d2500b2039.png)

**Comment by onecrayon** [comment]
@ycd Per [my comment here](https://github.com/tiangolo/fastapi/issues/1999#issuecomment-686228362), that's what I'm doing. It's also unnecessary code duplication. My feature request is for an easy way to pass in children of HTTPException and just have FastAPI automatically figure out what the error output should look like. So for example here's some pseudo-code that's based on my current project:

```python
from fastapi import status, HTTPException

class APIException(HTTPException):
    """Light wrapper around HTTPException that allows specifying defaults via class property"""

    status_code = status.HTTP_400_BAD_REQUEST
    detail = None
    headers = None

    def __init__(self, *args, **kwargs):
        if "status_code" not in kwargs:
            kwargs["status_code"] = self.status_code
        if "detail" not in kwargs:
            kwargs["detail"] = self.detail
        if "headers" not in kwargs:
            kwargs["headers"] = self.headers
        super().__init__(*args, **kwargs)

class TeapotException(APIException):
    status_code = 418
    detail = "I'm a teapot."

@app.get("/")
def read_root(gimme_coffee: bool = False, response_exceptions=[TeapotException]):
    if gimme_coffee:
        raise TeapotException()
    return {"Hello": "World"}
```

Even better would be to have FastAPI able to introspect the method and look for HTTPException-derived exceptions to auto-populate the list, but that's possibly edging a little too far into magical territory.

In any case, it's silly that I can have FastAPI automatically discover query params and such based on passed classes, but need to explicitly duplicate all of the information that's already stored in my exception classes in order to document exceptions.

**Comment by skaanbilly2** [comment]
Is this issue on the planning, or can I create a PR for this?

**Comment by Kludex** [comment]
Does this helps? https://github.com/Kludex/fastapi-responses

**Comment by onecrayon** [comment]
@Kludex That is *exactly* the sort of thing that I'm looking for! Granted, it's a little more "magical" than I usually like, but I would happily take it over the current setup where I have to explicitly define status codes and error strings multiple times.

**Comment by jtv8** [comment]
@onecrayon - given that the main use case for this is setting some expectations for API users on what HTTP responses to handle, would this suffice?

```
@app.get("/", response_exceptions=[418])
def read_root(gimme_coffee: bool = False):
    if gimme_coffee:
        raise HTTPException(status_code=418, detail="I'm a teapot.")
    return {"Hello": "World"}
```

**Comment by onecrayon** [comment]
@jtv8 No, that's insufficient. The API spec needs to be able to output not only the status code, but also the specification for the response body (since different exceptions often have different JSON body responses that get rendered, or other methods for reporting what went wrong).

**Comment by aiguofer** [comment]
+1 here. I'm currently designing various exceptions, each having a slightly different response payloads. I would love to have a convenient way to specify those different exceptions in our swagger docs. I like @onecrayon's example usage... it seems pretty consistent with existing FastAPI paradigms. However, in my case I'm also modifying the payload, so I'm not sure how that could be handled.

In my case:

```
class SessionConflictException(HTTPException):
    def __init__(self, session_id) -> None:
        self.session_id = session_id
        super().__init__(
            status_code=status.HTTP_409_CONFLICT,
            detail="Conflict: Session already exists and has not been terminated",
        )

@app.exception_handler(SessionConflictException)
async def session_conflict_exception_handler(
    request: Request, exc: SessionConflictException
):
    return JSONResponse(
        content=jsonable_encoder({"detail": exc.detail, "session_id": exc.session_id}),
        status_code=exc.status_code,
    )
```

I feel like this could be achieved somehow combining a pydantic `BaseModel` and a fastapi `HTTPException` into some sort of base class, with the only requirement being that it must contain a valid `status_code`.

**Comment by thedmdim** [comment]
Any update here?

**Comment by leehanchung** [comment]
Bump. In the absence of any decisions, what is the proper pattern for documenting the HTTPExceptions?

**Comment by jackirvine97** [comment]
Have there been any advancements on most DRY compliant pattern for ensuring custom HTTPExceptions are displayed in OpenAPI docs?

**Comment by JoakimJoensuu** [comment]
What I would like to see is using Pydantic model to define contents of error response body. 

In my case I have client and server that share common library which contains all request and response bodies defined as Pydantic models. This helps in interpreting req&resp bodies on both sides.

**Comment by yovelcohen** [comment]
Just for anyone passing by, here's how I'm handling it, I don't think it's ideal, but does the trick.
It creates a pydantic model dynamically with the given message and error_name.
you can create any error, but since I have a lot of crud APIs I chose to add the two extra utilities that create error by passing on the original response model your endpoint returns.

```python
def APIExcpetion(error_name, msg):
    error_name += 'Error'
    return create_model(error_name, detail=(str, msg))

def ObjNotFoundError(obj: Type[BaseModel], lookup_field=None):
    name = obj.__name__
    msg = f'{name} with {lookup_field} not found' if lookup_field else f'{name} not found'
    return APIExcpetion('NotFound', msg)

def ObjExistError(obj, lookup_field=None):
    name= obj.__name__
    msg = f'{name} with {lookup_field} already exists' if lookup_field else f'{name} already exists'
    return APIExcpetion('ObjectExist', msg)
```

Example Usage:
```python
@router.post(
    '/{client_id}/channels',
    response_model=ClientChannel,
    responses={404: {'model': ObjNotFoundError(Client, 'id')},
               400: {'model': ObjExistError(ClientChannel, 'name')}}
)
async def create_channel(
        user: Annotated[User, Depends(get_current_glix_user)],
        client_id: PydanticObjectId,
        channel_form: ClientChannelForm
) -> ClientChannel:
    """
    Create a new channel for a client
    """
    # I'm using Beanie, so this is just some pseudo functions to get your models...
    client, channel_exist = await asyncio.gather(get_client(client_id), check_channel_exists(client_id,channel_form.name))
    if not client:
        raise HTTPException(status_code=404, detail='Client not found')
    if channel_exist:
        raise HTTPException(status_code=400, detail='Channel with this name and client already exists')
    # do stuff and return model...
```

**Comment by jeremyschiemann** [comment]
+1

For me it would be enough to have extra responses coupled to a dependency, so that when an endpoint is using the dependency, it automatically adds the response to it. (see https://github.com/tiangolo/fastapi/discussions/11144)

**Comment by estebanx64** [comment, marked as answer]
Hi there.

As many others have mentioned, the only way so far is following this section from the doc [additional-responses](https://fastapi.tiangolo.com/advanced/additional-responses/)

or even including the responses at a router level.
```python
app.include_router(
    router,
    responses={
        418: {
            "description": "I'm a teapot",
            "content": {"application/json": {"example": {"dummy": "dumdum"}}},
        },
        413: {
            "description": "I'm too large",
            "content": {"application/json": {"example": {"dummy": "dumdum"}}},
        },
    },
)
```

I'm marking this question as answered and labeling as possible feature. We don't have any due date or whether is gonna be included or not yet

**Comment by PlugNPush** [comment]
Correct me if I am wrong, but out of all the attempts described here, there is no way to properly document two identical HTTP status codes with different "example" messages?

**Reply by Rey092** [reply]
> Correct me if I am wrong, but out of all the attempts described here, there is no way to properly document two identical HTTP status codes with different "example" messages?

or just use "examples": [] instead of "example": {}

**Comment by Rey092** [comment]
https://pypi.org/project/fastapi_errors/
you can try my package. multiple examples for one error code is supported
builded it in a few hours so be careful or just copy paste source to update by yourself if needed

**Comment by ale18V** [comment]
After reading [this discussion](https://github.com/pydantic/pydantic/issues/1875), I have come up with the following solution:
```python
@dataclass
class APIErrorModel:
    detail: str


class APIError(HTTPException, APIErrorModel):
    ...
```
It is less elegant than using  `HTTPException` directly as a model but I think it's better than other solutions I've seen.

Here is how it can be used:
```python
@router.get("/exc", responses={409: {"model": APIError}})
def rais():
    raise APIError(409, detail="bar")
```

Resulting openAPI:
![image](https://github.com/user-attachments/assets/5399dec3-09b0-4ae1-84b4-1bfba8546697)

**Comment by canton7** [comment]
I know this is old, but it popped up in my googling.

I ended up taking the following approach:

```python
ResponseType = dict[int | str, dict[str, Any]] | None


class GenericErrorResponse(BaseModel):
    """A simple generic error response model."""

    detail: str


class JSONModelResponse(Response):
    """A variant on JSONResponse which allows any Pydantic BaseModel as its content."""

    media_type = "application/json"

    def __init__(
        self,
        content: BaseModel,
        status_code: int = 200,
        headers: Mapping[str, str] | None = None,
        media_type: str | None = None,
        background: BackgroundTask | None = None,
    ) -> None:
        super().__init__(content, status_code, headers, media_type, background)

    def render(self, content: BaseModel) -> bytes:
        return content.model_dump_json().encode("utf-8")


class ErrorHandlingRoute(APIRoute):
    """
    Custom APIRoute subclass to be used with handles_exception.

    This must be configured on your router, e.g. 'app.router.route_class = ErrorHandlingRoute.
    """

    def __init__(
        self,
        path: str,
        endpoint: Callable[..., Any],
        responses: ResponseType = None,
        **kwargs: Any,
    ) -> None:
        # Pull any extra responses which were added to the endpoint by handles_exception
        extra_responses = cast(ResponseType, getattr(endpoint, "extra_responses", None))
        r = None if endpoint is None and extra_responses is None else (responses or {}) | (extra_responses or {})
        super().__init__(path, endpoint, responses=r, **kwargs)


def handles_exception[E: Exception, R: BaseModel](
    handler_or_exc_type: Callable[[E], R] | type[E],
    *,
    status_code: int,
    description: str,
) -> Callable[[DecoratedCallable], DecoratedCallable]:
    """
    Mark a route as handling a particular exception type, returning a given status code and description.

    Add this AFTER the @app.get/etc, e.g.

        @app.get("/")
        @handles_exception(some_handler, status=code=401, description="Some Description")
        def route(...):

    Where some_handler is a function which takes an Exception and returns a Pydantic BaseModel:

        def some_handler(exc: SomeException) -> SomeModel:

    This decorator reflects over the handler to extract the exception and model types, adds this so the OpenAPI
    extra_responses, and wraps the route in an exception handler which returns the serialized Pydantic model.

    Instead of a handler you can also pass an exception type, which will be serialized as a GenericErrorResponse

    Note that you must be using ErrorHandlingRoute.
    """

    def default_handler(exc: Exception) -> GenericErrorResponse:
        return GenericErrorResponse(detail=str(exc))

    handler: Callable[[Exception], BaseModel]
    exc_type: type[Exception]

    if isinstance(handler_or_exc_type, type):
        handler = default_handler
        exc_type = handler_or_exc_type
        model_type = GenericErrorResponse
    else:
        handler = cast(Callable[[Exception], BaseModel], handler_or_exc_type)
        hints = get_type_hints(handler)
        exc_type = next(iter(hints.values()))
        model_type = hints["return"]

    def decorator(endpoint: DecoratedCallable) -> DecoratedCallable:
        @functools.wraps(endpoint)
        async def wrapper(*args: Any, **kwargs: Any) -> Any:
            try:
                return await endpoint(*args, **kwargs)
            except exc_type as e:
                return JSONModelResponse(
                    status_code=status_code,
                    content=handler(e),
                )

        # Add this as an extra responses to the endpoint, to be extracted by ErrorHandlingRoute
        extra_responses = cast(dict[int | str, dict[str, Any]], getattr(endpoint, "extra_responses", {}))
        wrapper.extra_responses = extra_responses | {status_code: {"model": model_type, "description": description}}  # type: ignore
        return cast(DecoratedCallable, wrapper)

    return decorator
```

With:

```python
app.router.route_class = ErrorHandlingRoute
```

Then to get a simple response with "detail" you can do:

```python
class MyError(Exception):
    pass

@app.get("/")
@handles_exception(MyError, status_code=415, description="Some description")
def route():
    raise MyError("Some Message")
```

And to get a more structured response you can do:

```python
class MyErrorResponse(BaseModel):
    field_a: str
    field_b: str

def handle_my_error(exc: MyError) -> MyErrorResponse:
    return MyErrorResponse(...)

@app.get("/")
@handles_exception(handle_my_error, status_code=415, description="Some description")
def route():
    raise MyError("Some Message")
```
